pdc/apps/utils/messaging.py

```diff
@@ -179,21 +185,23 @@ def send_messages(self, msgs):
         This can save a lot of time for bulk changes since there will be only
         one connection per group.
         """
-        for group in itertools.groupby(msgs, lambda x: x[0]):
-            self._send_group(*group)
+        for topic, msgs in itertools.groupby(msgs, lambda x: x[0]):
+            self.queue.put((topic, list(msgs)))
```

Comment:
Can you call Queue.join() when httpd is being stopped (e.g. before upgrade)?

Reply:
I'm not sure, there does not seem to be any signal or hook that could be used for that. Is it useful though? What would it achieve?

I'm wondering whether using `multiprocessing` would be better than the thread.

Reply:
I'm not sure what happens with the thread when httpd is stopped.

Reply:
I would expect the main thread that runs the application to be stopped, and since this thread is marked as `daemon`, it should be killed (resulting in loss of any unsent messages still in queue).

Reply:
Well, there could be also some network issues so the reliability of actually delivering the messages could be resolved later if needed (using RabbitMQ would be nice but will probably take a while to implement with fixing all ansible scripts).